cohortextractor/dsl.py

```diff
@@ -0,0 +1,309 @@
+
+    >>> e = tables.clinical_events
+    >>> series = e.sort_by(e.date).first_for_patient().select_column(e.code)
+    >>> for item in series._path:
+            print(item)
+    (<concepts.tables.ClinicalEvents object at 0x7f211195fbb0>, 'sort_by', ('date',), {})
+    (<dsl.SortedEventFrame object at 0x7f211198d5b0>, 'first_for_patient', (), {})
+    (<dsl.PatientFrame object at 0x7f211198d610>, 'select_column', ('code',), {})
+
+When we compile a PatientSeries into a query_language.Value, we iterate over the path to
+build the appropriate Query Model objects.
+
+For each public DSL method on the DSL classes, there is a corresponding "private"
+compilation method that gets called in compilation to build an Query Model object.
+
+For instance, PatientFrame.select_column returns a PatientSeries, while
+PatientFrame._select_column is passed a query_language.Row, and calls .get on it to
+return a query_language.ValueFromRow.
+
+In some cases, we are not able to build the appropriate Query Model objects straight
+away.  An example of this is EventFrame._sort_by: in the Query Model, sorting happens in
+BaseTable.first_by, but when _sort_by is called, we don't know whether the user has
+requested the first or the last row for each patient.  To handle this, each compilation
+function can set some state which is passed to subsequent compilation functions, and in
+this example, _sort_by adds sort_columns to the state.
+
+Other notes...
+
+To support providing helpful error messages, we can implement __getattr__ on each class.
+This will intercept any lookup of a missing attribute, so that if eg a user tries to
+select a column from a SortedEventFrame, we can tell them they need to aggregate the
+SortedEventFrame first_for_patient.
+
+This docstring, and the function docstrings in this module are not currently intended
+for end users.
+"""
+
+from __future__ import annotations
+
+from typing import Any
+
+from cohortextractor.query_language import BaseTable as QMTable
+from cohortextractor.query_language import (
+    QueryNode,
+    Row,
+    Table,
+    Value,
+    ValueFromAggregate,
+    ValueFromRow,
+)
+
+
+class Cohort:
+    """Represents the cohort of patients in a study."""
+
+    def set_population(self, population: PatientSeries) -> None:
+        """Set the population variable for this cohort."""
+
+    def add_variable(self, name: str, variable: PatientSeries) -> None:
+        """Add a variable to this cohort by name."""
+        object.__setattr__(self, name, variable)
+
+    def __setattr__(self, name: str, variable: PatientSeries) -> None:
+        return self.add_variable(name, variable)
+
+
+class Node:
+    name: str
+
+    def __init__(
+        self,
+        parent: Node,
+        method_name: str,
+        *args: Any,
+        **kwargs: Any,
+    ):
+        self._path: list[PathItem] = parent._path + [
+            (parent, method_name, args, kwargs)
+        ]
+
+
+PathItem = tuple[Node, str, Any, Any]
+State = dict[str, Any]
```

Comment:
With the current use of State, it looks like it could be `dict[str, str]`, are you planning for there to be any other types of value in extra_state?

Reply:
> Could these Anys be tightened at all? Even just to object?

I don't think so -- if we use object, then mypy will object if we try to do anything with the objects.

There's an invariant here that is going to be very hard to encode with types: the compilation functions only pull things out of state if we know that the values are going to be there.

While this could be `dict[str, str]` for now, I think I'm going to leave this as is too.